### ISSUE TITLE
Better JS debugging experience (2)

### ISSUE THREAD

**Issue by ncannasse** [HaxeFoundation/haxe#9327]
Following #9035, there's still a few things to take care of I think:

- temp var name generation : I'm not sure if the patch addressed that, the idea is to reduce the amount of _g1,_g2 etc generated in output code. Adding variable name mapping in sourcemaps would be even better since we could either remove them entirely or collapse them into a single "<temp>" name.

- stepping : when stepping with F10, we often stay on the same Haxe line and have to step several times to go to the next line. I'm not sure if that's a problem with sourcemaps or if we need to change JS output so one line of Haxe gets generated within the same JS line.

**Comment by RealyUniqueName** [comment]
> temp var name generation : I'm not sure if the patch addressed that

The patch does reduce unique temp var names to some extent. Since Haxe tries to generate temp vars only when absolutely needed  (with analyzer enabled), I doubt we can reduce amount of temp vars significantly.
Name mapping might be an option. I'll investigate it.

> stepping : when stepping with F10, we often stay on the same Haxe line and have to step several times to go to the next line.

This is caused by Haxe expressions being compiled into multiple expressions or even multiple lines of expression of a target language. 
I wonder if it will help to generate a single source map entry for such multi-liners instead of an entry per a JS expression.

**Comment by ncannasse** [comment]
@RealyUniqueName I was thinking about temp vars such as switch value, or loop iterators.
Here's some output with latest Haxe (breakpoint set in Hide hide.cdb.comp.Cell.valueHtml inside double "for" loop - case List)

Without ES6:
![image](https://user-images.githubusercontent.com/1022912/79794637-e6be2780-8352-11ea-834a-7b8d0dd7ce6d.png)


With ES6, it's a lot better, thanks to let, but debugger creates a lot of "Block" that needs expending, I don't like it very much (not our fault...)
![image](https://user-images.githubusercontent.com/1022912/79794534-b5ddf280-8352-11ea-9f8f-7be2423cefd4.png)

**Comment by RealyUniqueName** [comment]
Yeah, because of var hoisting we can't re-use the same name for var shadowing inside of loops. If name mapping via sourcemaps won't help, then I don't think we can do anything else.

**Comment by Simn** [comment]
Is there anything actionable left to do here after the recent local renaming changes?

**Comment by RealyUniqueName** [comment]
We still need to explore name-mapping in source maps
